[GTK][WPE] Build removing unused sections when supported

magomez opened the pull request:
#### 8603570132a8d1abf94583c2c695b53e83e018e9
<pre>
[GTK][WPE] Build removing unused sections when supported
<a href="https://bugs.webkit.org/show_bug.cgi?id=270902">https://bugs.webkit.org/show_bug.cgi?id=270902</a>

Reviewed by Adrian Perez de Castro.

Improve the binary size by removing unused sections when supported
by the compiler and the linker.

* Source/cmake/OptionsCommon.cmake:

Canonical link: <a href="https://commits.webkit.org/276726@main">https://commits.webkit.org/276726@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fe97c7daa8f9cb476ab41693696dace201321f24

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/45370 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/24492 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/47897 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/48034 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/41378 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/28716 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/21887 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/37213 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/45948 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/21565 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/39136 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/18315 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/18974 "Passed tests") | | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/3417 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/38591 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/41657 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/40538 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/49754 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/44839 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/20354 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/16875 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/44262 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/21660 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/43084 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10114 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/22020 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/51998 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/21348 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/10625 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Source/cmake/OptionsWPE.cmake:
```diff
@@ -452,5 +453,26 @@ else ()
+        set(CMAKE_C_FLAGS_DEBUG "${CMAKE_C_FLAGS_DEBUG}${CMAKE_COMPILER_SIZE_OPT_FLAGS} -ffunction-sections -fdata-sections")
+        set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG}${CMAKE_COMPILER_SIZE_OPT_FLAGS} -ffunction-sections -fdata-sections")
+        set(CMAKE_SHARED_LINKER_FLAGS_DEBUG "${CMAKE_SHARED_LINKER_FLAGS_DEBUG} -Wl,--gc-sections")
+    endif ()
```
You could replace all this with a single `WEBKIT_APPEND_GLOBAL_COMPILER_FLAGS` invocation, except for the linker flags. For the linker flags I would add a `LD_SUPPORTS_GC_SECTIONS` variable in `OptionsCommon.cmake` (we already have some `LD_SUPPORTS_<option>` there), and skip using the variables with the `_RELEASE` and `_DEBUG` suffixes to use the flags in all kinds of builds. Lastly, I would also set the executable and loadable shared module flags.

Like this:

```suggestion
    WEBKIT_APPEND_GLOBAL_COMPILER_FLAGS(-finline-limit=90
                                        -fsection-anchors)
    if (LD_SUPPORTS_GC_SECTIONS)
        WEBKIT_APPEND_GLOBAL_COMPILER_FLAGS(-ffunction-sections
                                            -fdata-sections)
        set(CMAKE_EXE_LINK_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -Wl,--gc-sections")
        set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} -Wl,--gc-sections")
        set(CMAKE_MODULE_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} -Wl,--gc-sections")
    endif ()
```

Like @mcatanzaro mentioned, this should probably go in `WebKitCompilerFlags.cmake` or in `OptionsCommon.cmake`.

I think we are reaching a point of linker flag tweaking that would justify having helper CMake functions to check whether the linker supports a given flag, and another to pass a list of flags and automatically add the ones supported by the linker. But that could be a follow-up patch.

aperezdc replied:
FWIW, I think it does _not_ make sense to apply the flags only to `MinSizeRel` builds. One may want an optimized build and at the same time do the `--gc-sections` part&mdash;because it always provides a gain.

OTOH, options `-finline-limit=` and `-fsection-anchors` may produce less performant code, so those we would want only for `MinSizeRel` builds so these two makes sense to add to `CMAKE_{C,CXX}_FLAGS_MINSIZEREL`. Recently I did a refactor in our CMake compiler-flag functions which allows doing the same:

```cmake
WEBKIT_VAR_ADD_COMPILER_FLAGS(C APPEND CMAKE_C_FLAGS_MINSIZEREL
    -finline-limit=90 -fsection-anchors)
WEBKIT_VAR_ADD_COMPILER_FLAGS(CXX APPEND CMAKE_CXX_FLAGS_MINSIZEREL
    -finline-limit=90 -fsection-anchors)
```

Only those options that are supported by the compiler will be appended to the variables.

aperezdc reviewed (approved):
@magomez Good idea reducing the scope, now I think the change in this patch should not be controversial 👍🏼 

I would still like to have a CMake option to enable a set of “recommended optimizations” that would be enabled by default but could be opted-out from, to be used by packagers which may prefer to apply the compiler flags in other ways &mdash; but let's leave this as-is and do it as a follow-up.

magomez committed:
[GTK][WPE] Build removing unused sections when supported
https://bugs.webkit.org/show_bug.cgi?id=270902

Reviewed by Adrian Perez de Castro.

Improve the binary size by removing unused sections when supported
by the compiler and the linker.

* Source/cmake/OptionsCommon.cmake:

Canonical link: https://commits.webkit.org/276726@main
